cos_alerter/alerter.py

```diff
@@ -29,24 +34,67 @@ def __exit__(self, _, __, ___):
         fcntl.lockf(self.fh, fcntl.LOCK_UN)
         self.fh.close()
 
-    @property
-    def alert_time(self):
-        return self.data['alert_time']
+    @staticmethod
+    def initialize():
+        # Note this method does not do any locking so be responsible.
+        current_date = datetime.datetime.now(datetime.timezone.utc)
+        current_time = time.monotonic()
+        data = {
+            # The actual date and time that COS Alerter was started.
+            'start_date': datetime.datetime.timestamp(current_date),
 
-    @alert_time.setter
-    def alert_time(self, value):
-        self.data['alert_time'] = value
+            # The time according to the monotonic clock that COS Alerter was started.
+            'start_time': current_time,
 
-    @property
-    def notify_time(self):
-        return self.data['notify_time']
+            # The last time we received an alert from Alertmanager.
+            'alert_time': current_time,
 
-    @notify_time.setter
-    def notify_time(self, value):
-        self.data['notify_time'] = value
+            # The last time we sent out notifications.
+            'notify_time': None,
+        }
+
+        with open(config['watch']['data_file'], 'w') as f:
+            json.dump(data, f)
+
+    def set_alert_time(self):
+        self.data['alert_time'] = time.monotonic()
+
+    def _set_notify_time(self):
+        self.data['notify_time'] = time.monotonic()
+
+    def is_down(self):
+        down_interval = durationpy.from_str(config['watch']['down_interval']).total_seconds()
+        return time.monotonic() - self.data['alert_time'] > down_interval
+
+    def _recently_notified(self):
+        repeat_interval = durationpy.from_str(config['notify']['repeat_interval']).total_seconds()
+        return self.data['notify_time'] and not time.monotonic() - self.data['notify_time'] > repeat_interval
+
+    def _last_alert_datetime(self):
+        actual_alert_timestamp = (self.data['alert_time'] - self.data['start_time']) + self.data['start_date']
+        return datetime.datetime.fromtimestamp(actual_alert_timestamp, datetime.timezone.utc)
+
+    def notify(self):
+        # If we have already notified recently, do nothing.
+        if self._recently_notified():
+            return
+
+        self._set_notify_time()
+        last_alert_time = self._last_alert_datetime().isoformat()
+        title = '**Alertmanager is Down!**'
```

Comment:
for better testability, I'd extract this to a module-level function (or a staticmethod, also good). Something like `_notify_async`

Reply:
What do you mean by "this"? The entire notify method or just a part?